x/valset-pref/msg_server_test.go

```diff
@@ -124,3 +124,82 @@ func (suite *KeeperTestSuite) TestSetValidatorSetPreference() {
+	preferences := suite.PrepareDelegateToValidatorSet()
+
+	tests := []struct {
+		name           string
+		delegator      sdk.AccAddress
+		coin           sdk.Coin
+		expectedShares []sdk.Dec
+		expectPass     bool
+		valSetExists   bool
+	}{
+		{
+			name:           "Delegate to valid validators!",
+			delegator:      sdk.AccAddress([]byte("addr1---------------")),
+			coin:           sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(10_000_000)),                      // amount to delegate
+			expectedShares: []sdk.Dec{sdk.NewDec(5_000_000), sdk.NewDec(3_000_000), sdk.NewDec(2_000_000)}, // expected shares after delegation
+			expectPass:     true,
+		},
+		{
+			name:           "Delegate more tokens to existing validator-set",
+			delegator:      sdk.AccAddress([]byte("addr1---------------")),
+			coin:           sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(10_000_000)),                       // amount to delegate
+			expectedShares: []sdk.Dec{sdk.NewDec(10_000_000), sdk.NewDec(6_000_000), sdk.NewDec(4_000_000)}, // expected shares after delegation
+			expectPass:     true,
+			valSetExists:   true,
+		},
+		{
+			name:           "Delegate Decimal Amounts",
+			delegator:      sdk.AccAddress([]byte("addr2---------------")),
+			coin:           sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(25_000_000)),                       // amount to delegate
+			expectedShares: []sdk.Dec{sdk.NewDec(12_500_000), sdk.NewDec(7_500_000), sdk.NewDec(5_000_000)}, // expected shares after delegation
+			expectPass:     true,
+		},
+		{
+			name:       "User doesnot have enough tokens to stake",
+			delegator:  sdk.AccAddress([]byte("addr3---------------")),
+			coin:       sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(200_000_000)), // amount to delegate
+			expectPass: false,
+		},
+	}
+
+	for _, test := range tests {
+		suite.Run(test.name, func() {
+			suite.FundAcc(test.delegator, sdk.Coins{sdk.NewInt64Coin(sdk.DefaultBondDenom, 100_000_000)}) // 100 osmo
+
+			// setup message server
+			msgServer := valPref.NewMsgServerImpl(suite.App.ValidatorSetPreferenceKeeper)
+			c := sdk.WrapSDKContext(suite.Ctx)
+
+			if !test.valSetExists {
+				_, err := msgServer.SetValidatorSetPreference(c, types.NewMsgSetValidatorSetPreference(test.delegator, preferences))
+				suite.Require().NoError(err)
+			}
+
+			// call the create validator set preference
+			_, err := msgServer.DelegateToValidatorSet(c, types.NewMsgDelegateToValidatorSet(test.delegator, test.coin))
+			if test.expectPass {
+				suite.Require().NoError(err)
+
+				// check if the expectedShares matches after delegation
+				for i, val := range preferences {
+					valAddr, err := sdk.ValAddressFromBech32(val.ValOperAddress)
+					suite.Require().NoError(err)
+
+					//gurantees that the delegator exist because we check it in DelegateToValidatorSet
+					del, _ := suite.App.StakingKeeper.GetDelegation(suite.Ctx, test.delegator, valAddr)
+					suite.Require().Equal(del.Shares, test.expectedShares[i])
```

Comment:
Can we add a check on user balance and confirmed that it has been reduced as well?

Reply:
added [here](https://github.com/osmosis-labs/osmosis/pull/3260/commits/09ee056004a254b10660d332b42cdbce1d2e047a#)